### ISSUE TITLE
'runtime' not correctly parsed with --workflow-profile

### ISSUE THREAD

**Issue by cmeesters** [snakemake/snakemake#2696]
**Snakemake version**
8.4.8

**Describe the bug**
with a profile like

```
default-resources:
    slurm_partition: "smp"

set-resources:
    test1:
        slurm_partition: "parallel"
        slurm_extra: "'--nice=150'"
        runtime: 60
```

and a Snakefile 

```
rule all:
     input: "results/a.out"

rule test1:
     output: "results/a.out"
     shell: "touch {output}
```
using the SLURM executor, the executor receives a '1' for the runtime. Doubling the value to '120' translate to '2' - clearly a matter of human readable translation. Fine. Yet, changing to `runtime: '60m' yields:

```
InputFunctionException in rule test1 in file /gpfs/fs1/home/meesters/projects/hpc-jgu-lifescience/snakemake-workflows/test-workflow/Snakefile, line 4:
Error:
  WorkflowError:
    Failed to evaluate resources value '60m'.
        String arguments may need additional quoting. E.g.: --default-resources "tmpdir='/home/user/tmp'".
    SyntaxError: invalid decimal literal (<string>, line 1)
```

Placing this setting into the Snakefile will parse the runtime correctly, and 'runtime="60m"' will translate to "60". 

This is a little impractical: It imposes the necessity to define a workflow profile with definitions in seconds, for the submit plugin to use minutes.

**Comment by nikostr** [comment]
Maybe this is self-evident for people more familiar with the connection between the snakemake CLI and the workflow profiles, but this bug also affects setting the runtime as a default resource through the CLI using `--default-resources runtime='60m'`.

**Comment by cmeesters** [comment]
Yes, I think(!) it is. Thanks for pointing it out, anyway! 

Admittedly, I am not familiar with the parsing code. So this will be picked up or take a little longer ...

**Comment by nikostr** [comment]
I also noted now that specifying a number as a default resource in the workflow profile leads to that being used as-is - I might have read your first post wrong, but does this mean that runtime is being treated differently in `default-resources` and `set-resources`? That is, if I set runtime as 720 in the default resources, this gets set to SLURM as-is, and SLURM converts this from 720 minutes to 12 hours. Attempting to use human friendly syntax in the default resources ends up giving `WorkflowError: Cannot parse runtime value into minutes for setting runtime resource: <TBD>`.

**Comment by cmeesters** [comment]
- ok, when specified in the profile, `runtime` is of type <class 'snakemake.io.AnnotatedString'> and hence asking `isinstance(value, str)` becomes true. However, it may not be something like "50m", for the parsing occurs before that and the error is:
```
WorkflowError:
    Failed to evaluate resources value '50h'.
        String arguments may need additional quoting. E.g.: --default-resources "tmpdir='/home/user/tmp'".
    SyntaxError: invalid decimal literal (<string>, line 1)
```

- when specified in the Snakefile, the type is `int`.

So, we are dealing with two errors:

1. parsing one value type as two variable types, plus the difference in resources and default-resources. 
2. raising an error, which should not be raised because the string adheres to the required string format

Pew.

**Comment by blaiseli** [comment]
I noticed that when I specify `runtime: 0` for a rule in the `set-resources` section of the workflow-profile, the rule actually asks slurm for 1 minute.

I thought that this was supposed to ask for no time limit.

The `default-resources` section is set to `runtime: 120`, so if default was to be used instead of no limit, I would expect 2 minutes or 2 hours, depending on the current parsing style.

**Comment by cmeesters** [comment]
That is actually intentional, and the rational goes like this:

- a user may ask for any time, there is no check.
- a user may drop below a cluster default, there is no check, either. It's sensible because increased runtime, impacts fair share and if for a specific rule, a user may ask for less, that's fine.
- there is however a minimum: 0 or indefinitely might be an error, that would have a tremendous impact on a users' fair share and the throughput of a workflow. (As far as I know, 0 does not mean "take the default", if I try this on my cluster, I get `Unable to allocate resources: Requested time limit is invalid (missing or exceeds some limit)`.)

I know that there are some clusters without wall times. However, this is unusual, and I did not want users to suffer from such mistakes. You can still ask for weeks on end - if it makes you happy ;-).

I will update the executor docs accordingly.

**Comment by blaiseli** [comment]
Thanks for your explanations.

I'll try to reformulate to be sure I understood correctly.

1) I was mistaken when I thought asking for `runtime: 0` meant "no time limit".
2) If a rule has an explicit `runtime: 0`, the rule will instead ask 1
3) Otherwise, a rule uses the runtime defined in `default-resources`

Is that correct?

**Comment by cmeesters** [comment]
and

4. the `default-resources` can be overwritten on a per-rule basis in the workflow profile.

**Comment by cmeesters** [comment]
solved by the release of 8.5.0
